src/MoveUntilTouchTopicController.cpp

```diff
@@ -45,97 +34,54 @@ bool MoveUntilTouchTopicController::init(hardware_interface::RobotHW *robot,
   }
-  if (mForceLimit < 0) {
+  if (forceLimit < 0) {
     ROS_ERROR("sensor_force_limit must be positive or zero");
     return false;
   }
-  if (mTorqueLimit < 0) {
+  if (torqueLimit < 0) {
     ROS_ERROR("sensor_torque_limit must be positive or zero");
     return false;
   }
 
-  // subscribe to sensor data
-  mForceTorqueDataSub = nh.subscribe(
-      ft_wrench_name, 1,
-      &MoveUntilTouchTopicController::forceTorqueDataCallback, this);
-
-  // action client to kick off taring
-  mTareActionClient =
-      std::unique_ptr<TareActionClient>(new TareActionClient(nh, ft_tare_name));
-
-  // start action server
-  mFTThresholdActionServer.reset(
-      new actionlib::ActionServer<SetFTThresholdAction>{
-          nh, "set_forcetorque_threshold",
-          std::bind(&MoveUntilTouchTopicController::setForceTorqueThreshold,
-                    this, std::placeholders::_1),
-          false});
-  mFTThresholdActionServer->start();
+  // Init FT Threshold Server
+  mFTThresholdServer.reset(new FTThresholdServer{
+      nh, ft_wrench_name, ft_tare_name, forceLimit, torqueLimit});
 
   // initialize base trajectory controller
   return initController(robot, nh);
 }
 
-//=============================================================================
-void MoveUntilTouchTopicController::forceTorqueDataCallback(
-    const geometry_msgs::WrenchStamped &msg) {
-  std::lock_guard<std::mutex> lock(mForceTorqueDataMutex);
-  mForce.x() = msg.wrench.force.x;
-  mForce.y() = msg.wrench.force.y;
-  mForce.z() = msg.wrench.force.z;
-  mTorque.x() = msg.wrench.torque.x;
-  mTorque.y() = msg.wrench.torque.y;
-  mTorque.z() = msg.wrench.torque.z;
-  mTimeOfLastSensorDataReceived = std::chrono::steady_clock::now();
-}
-
-//=============================================================================
-void MoveUntilTouchTopicController::taringTransitionCallback(
-    const TareActionClient::GoalHandle &goalHandle) {
-  if (goalHandle.getResult() && goalHandle.getResult()->success) {
-    ROS_INFO("Taring completed!");
-    mTimeOfLastSensorDataReceived = std::chrono::steady_clock::now();
-    mTaringCompleted.store(true);
-  }
-}
-
 //=============================================================================
 void MoveUntilTouchTopicController::starting(const ros::Time &time) {
-  // start asynchronous tare request
-  ROS_INFO("Starting Taring");
-  pr_control_msgs::TriggerGoal goal;
-  mTareGoalHandle = mTareActionClient->sendGoal(
-      goal,
-      boost::bind(&MoveUntilTouchTopicController::taringTransitionCallback,
-                  this, _1));
-
   // start base trajectory controller
   startController(time);
+
+  // start FTThresholdServer
+  mFTThresholdServer->start();
 }
 
 //=============================================================================
 void MoveUntilTouchTopicController::stopping(const ros::Time &time) {
   // stop base trajectory controller
   stopController(time);
+
+  // stop FTThresholdServer
+  mFTThresholdServer->stop();
```

Comment:
Does the matter of stopping the controller and this server matter?

Reply:
Yeah, since we can switch between controllers in a given session (example: the Cartesian Velocity controller), and we don't want to have an unnecessary FTThresholdServer that is calling commands that could mess with the commands coming from the running controller, if that makes sense.

This effectively replaces the "isRunning()" check that you pointed out in the previous revision.